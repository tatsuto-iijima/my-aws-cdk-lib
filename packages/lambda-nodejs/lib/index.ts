import * as cdk from '@aws-cdk/core';
import * as lambdaNodejs from '@aws-cdk/aws-lambda-nodejs';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as fs from 'fs';
const jsSHA = require("jssha");

export interface LambdaNodejsProps {
  func?: lambdaNodejs.NodejsFunctionProps,
  role?: {
    description?: string;
    managedPolicies?: string[];
  },
  version?: {
    removalPolicy: cdk.RemovalPolicy;
  };
}

export class LambdaNodejs extends cdk.Construct {
  public readonly role: iam.Role;

  public readonly nodejsFunction: lambdaNodejs.NodejsFunction;

  public readonly version: lambda.Version;

  constructor(scope: cdk.Construct, id: string, props: LambdaNodejsProps = {}) {
    super(scope, id);

    const finalLambdaNodejsProps: any = {};
    Object.assign(finalLambdaNodejsProps, props.func);
    
    if (props.role) {
      this.role = new iam.Role(this, 'Role', {
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        description: props.role.description,
      });
      if (props.role.managedPolicies) {
        for (const mangagedPolicie of props.role.managedPolicies) {
          this.role.addManagedPolicy(iam.ManagedPolicy.fromAwsManagedPolicyName(mangagedPolicie));
        }
      }
      finalLambdaNodejsProps.role = this.role;
    }

    this.nodejsFunction = new lambdaNodejs.NodejsFunction(this, 'Function', finalLambdaNodejsProps);

    if (props.version) {
      let entry = './lib/index.Function.ts'
      if (props.func?.entry) entry = props.func.entry;
      const shaObj = new jsSHA("SHA-256", "TEXT", { encoding: "UTF8" });
      shaObj.update(fs.readFileSync(entry).toString());
      const hash = shaObj.getHash("HEX");
      this.version = new lambda.Version(this, 'Version-' + hash, {
        lambda: this.nodejsFunction,
        removalPolicy: props.version.removalPolicy,
      });
    }

  }

}
