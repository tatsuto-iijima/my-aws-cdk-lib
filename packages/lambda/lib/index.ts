import * as cdk from '@aws-cdk/core';
import * as iam from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';

export interface LambdaProps {
  func: lambda.FunctionProps;
  role?: {
    servicePrincipalNames: string[];
    description?: string;
    managedPolicieNames?: string[];
  };
  version?: {
    name: string;
    removalPolicy?: cdk.RemovalPolicy;
  };
}

export class Lambda extends cdk.Construct {
  public readonly role: iam.Role;

  public readonly func: lambda.Function;

  public readonly version: lambda.Version;

  constructor(scope: cdk.Construct, id: string, props: LambdaProps) {
    super(scope, id);

    const finalLambdaProps: any = {};
    Object.assign(finalLambdaProps, props.func);
    
    if (!finalLambdaProps.role && props.role) {
      const finalRoleProps:any = {};
      const servicePrincipals: iam.ServicePrincipal[] = [];
      for (const servicePricipalName of props.role.servicePrincipalNames) {
        servicePrincipals.push(new iam.ServicePrincipal(servicePricipalName));
      }
      finalRoleProps.assumedBy = new iam.CompositePrincipal(...servicePrincipals);
      if (props.role.description) finalRoleProps.description = props.role.description;
      if (props.role.managedPolicieNames) {
        const managedPolicies: iam.IManagedPolicy[] = [];
        for (const mangagedPolicieName of props.role.managedPolicieNames) {
          managedPolicies.push(iam.ManagedPolicy.fromAwsManagedPolicyName(mangagedPolicieName));
        }
        finalRoleProps.managedPolicies = managedPolicies;
      }
      this.role = new iam.Role(this, 'Role', finalRoleProps);
      finalLambdaProps.role = this.role;
    }

    this.func = new lambda.Function(this, 'Function', finalLambdaProps);

    if (props.version) {
      const finalVersionProps:any = {
        lambda: this.func,
      };
      if (props.version.removalPolicy) finalVersionProps.removalPolicy = props.version.removalPolicy;
      this.version = new lambda.Version(this, 'Version-' + props.version.name, finalVersionProps);
    }

  }

}
