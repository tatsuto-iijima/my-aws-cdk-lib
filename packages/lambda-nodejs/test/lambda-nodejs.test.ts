import { expect as expectCDK, countResources, haveResource, ABSENT, arrayWith, objectLike, haveResourceLike, SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as AwsCdkLibLambdaNodejs from '../lib/index';

describe('Test Construct "LambdaNodejs" by default', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  new AwsCdkLibLambdaNodejs.LambdaNodejs(stack, 'MyTestConstruct');

  test('Lambda Function x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1));
  });

  // In aws-lambda-nodejs, "lambda.amazonaws.com" is set in AssumeRole
  // and "AWSLambdaBasicExecutionRole" is set in ManagedPolicy　as standard Role.
  test('IAM Role x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::IAM::Role", 1));
  });
  test('IAM Role Property "AssumeRolePolicyDocument" has "lambda.amazonaws.com"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: arrayWith(
          objectLike({
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: "lambda.amazonaws.com",
            },
          }),
        ),
      },
    }));
  });
  test('IAM Role Property "ManagedPolicyArns" has "AWSLambdaBasicExecutionRole"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      ManagedPolicyArns: arrayWith(objectLike({
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          ],
        ]
      })),
    }));
  });

  test('Lambda Version x0 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Version", 0));
  });

  // test('Snapshot', () => {
  //   expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  // });
});

const testRolePropsData = [
  {},
  { description: 'Test Construct "LambdaNodejs"' },
  { managedPolicies: [ 'service-role/AWSLambdaBasicExecutionRole' ] },
  { description: 'Test Construct "LambdaNodejs"' , managedPolicies: [ 'service-role/AWSLambdaBasicExecutionRole' ] },
  { description: 'Test Construct "LambdaNodejs"' , managedPolicies: [ 'service-role/AWSLambdaBasicExecutionRole', 'service-role/AmazonS3ReadOnlyAccess' ] },
];
const testRolePropsTable = [];
for (const testRoleProps of testRolePropsData) {
  testRolePropsTable.push([
    JSON.stringify(testRoleProps),
    testRoleProps,
  ]);
}

// console.log(JSON.stringify(testRolePropsTable, undefined, 2));  // for Debug

describe.each(testRolePropsTable)('Test Construct "LambdaNodejs" by option { role: %j }', (json, testRoleProps) => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  if (typeof testRoleProps != 'string') {
    new AwsCdkLibLambdaNodejs.LambdaNodejs(stack, 'MyTestConstruct', {
      role: testRoleProps,
    });
  } else {
    throw new Error('ERROR: Type of RoleProps is string');
  }
  
  test('Lambda Function x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1));
  });
  
  test('IAM Role x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::IAM::Role", 1));
  });

  test('IAM Role Property "AssumeRolePolicyDocument" has "lambda.amazonaws.com"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: arrayWith(
          objectLike({
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: "lambda.amazonaws.com",
            },
          }),
        ),
      },
    }));
  });

  if (testRoleProps.description) {
    test('IAM Role Property "Description" is "' + testRoleProps.description + '"', () => {
      expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
        Description: testRoleProps.description,
      }));
    });
  } else {
    test('IAM Role Property "Description" is none', () => {
      expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
        Description: ABSENT,
      }));
    });
  }

  if (testRoleProps.managedPolicies) {
    test.each(testRoleProps.managedPolicies)('IAM Role Property "ManagedPolicyArns" is "%s"', (managedPoliciesName) => {
      expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
        ManagedPolicyArns: arrayWith(objectLike({
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::aws:policy/" + managedPoliciesName,
            ],
          ]
        })),
      }));
    });
  } else {
    test('IAM Role Property "ManagedPolicyArns" is none', () => {
      expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
        ManagedPolicyArns: ABSENT,
      }));
    });
  }

  test('Lambda Version x0 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Version", 0));
  });

});

describe.each([
  [ { version: { removalPolicy: cdk.RemovalPolicy.DESTROY } } ],
  [ { version: { removalPolicy: cdk.RemovalPolicy.RETAIN } } ],
  [ { func: { entry: './lambda/test/index.ts', handler: 'handler' }, version: { removalPolicy: cdk.RemovalPolicy.DESTROY } } ],
])('Test Construct "LambdaNodejs" by option %j', (option) => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  new AwsCdkLibLambdaNodejs.LambdaNodejs(stack, 'MyTestConstruct', option);

  test('Lambda Function x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Function", 1));
  });

  // In aws-lambda-nodejs, "lambda.amazonaws.com" is set in AssumeRole
  // and "AWSLambdaBasicExecutionRole" is set in ManagedPolicy　as standard Role.
  test('IAM Role x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::IAM::Role", 1));
  });
  test('IAM Role Property "AssumeRolePolicyDocument" has "lambda.amazonaws.com"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Statement: arrayWith(
          objectLike({
            Action: "sts:AssumeRole",
            Effect: "Allow",
            Principal: {
              Service: "lambda.amazonaws.com",
            },
          }),
        ),
      },
    }));
  });
  test('IAM Role Property "ManagedPolicyArns" has "AWSLambdaBasicExecutionRole"', () => {
    expectCDK(stack).to(haveResourceLike("AWS::IAM::Role", {
      ManagedPolicyArns: arrayWith(objectLike({
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          ],
        ]
      })),
    }));
  });
  
  test('Lambda Version x1 Created', () => {
    expectCDK(stack).to(countResources("AWS::Lambda::Version", 1));
  });
  
});

const option =  {
  func: {
    entry: './lambda/test/index.ts',
    handler: 'handler',
  },
  role: {
    description: "for MyAwsCdkLibLambdaNodejsTest2Stack",
    managedPolicies: [
      "service-role/AWSLambdaBasicExecutionRole",
      "AmazonS3ReadOnlyAccess"
    ],
  },
  version: {
    removalPolicy: cdk.RemovalPolicy.RETAIN,
  },
};
describe('Test Construct "LambdaNodejs" by option ' + JSON.stringify(option), () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, "TestStack");
  new AwsCdkLibLambdaNodejs.LambdaNodejs(stack, 'MyTestConstruct', option);

  // test('Snapshot', () => {
  //   expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  // });

});
