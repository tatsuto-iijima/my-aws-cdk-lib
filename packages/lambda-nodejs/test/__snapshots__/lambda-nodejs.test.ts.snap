// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Construct "LambdaNodejs" by default Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7ArtifactHashA2855EC6": Object {
      "Description": "Artifact hash for asset \\"4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7\\"",
      "Type": "String",
    },
    "AssetParameters4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7S3Bucket1946552D": Object {
      "Description": "S3 bucket for asset \\"4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7\\"",
      "Type": "String",
    },
    "AssetParameters4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7S3VersionKeyA13E0720": Object {
      "Description": "S3 key for asset version \\"4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyTestConstructFunction4795DE86": Object {
      "DependsOn": Array [
        "MyTestConstructFunctionServiceRole786CED66",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7S3Bucket1946552D",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7S3VersionKeyA13E0720",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4b7b725fb32b7e3c3a434db632c41812742b5fe3bbba92a400497764a9b174c7S3VersionKeyA13E0720",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructFunctionServiceRole786CED66",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructFunctionServiceRole786CED66": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`Test Construct "LambdaNodejs" by option {"func":{"entry":"./lambda/test/index.ts","handler":"handler"},"role":{"description":"for MyAwsCdkLibLambdaNodejsTest2Stack","managedPolicies":["service-role/AWSLambdaBasicExecutionRole","AmazonS3ReadOnlyAccess"]},"version":{"removalPolicy":"retain"}} Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ffArtifactHash6BE44837": Object {
      "Description": "Artifact hash for asset \\"23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ff\\"",
      "Type": "String",
    },
    "AssetParameters23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ffS3Bucket797C5718": Object {
      "Description": "S3 bucket for asset \\"23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ff\\"",
      "Type": "String",
    },
    "AssetParameters23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ffS3VersionKey568F954E": Object {
      "Description": "S3 key for asset version \\"23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ff\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "MyTestConstructFunction4795DE86": Object {
      "DependsOn": Array [
        "MyTestConstructRole16E7E65D",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ffS3Bucket797C5718",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ffS3VersionKey568F954E",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters23409060fae67c3d17b252ca4488da2090e3062e71f107f198a424e18704f9ffS3VersionKey568F954E",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTestConstructRole16E7E65D",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTestConstructRole16E7E65D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "for MyAwsCdkLibLambdaNodejsTest2Stack",
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/AmazonS3ReadOnlyAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTestConstructVersion4b4f2133c15e419ba344b8c5d9f61d919b29c38b7ba7bcd61e4bc01163d0f53dD881883A": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "MyTestConstructFunction4795DE86",
        },
      },
      "Type": "AWS::Lambda::Version",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
